THRUSTERS: fix timing dependent thruster detection

DSsoto opened the pull request:
### Thruster Comms
* Increase the serial port timeout from 0.01 s to 0.1 s.
* Parse the response to poll packets to confirm that the response packet
  is actually from the thruster we believe it to be.
* Instead of checking for about 100 thruster ID's on each port, only check the
  ones that have been declared on the thruster layout.
### ThrusterSpinner
* Changed allowable effort range to [-1, 1] so we can spin in both directions
* Added a thruster discovery mode to the thruster spinner utility so that we can
  find out what thrusters are plugged into a port even if they were not declared
  on the thruster_layout.yaml
* Added reporting average packet turnaround time to discovery mode.
**_Proof:_** 
I swapped the connectors for the BR and FR ports on the back of the sub. I ran
the thruster spinner utility and was told that half of the thrusters were missing.
Then I entered discovery mode by pressing 'C' and was able to find the missing
thrusters.
![image](https://cloud.githubusercontent.com/assets/8714358/25777694/c87f6502-32b2-11e7-958f-53758d3f4f9c.png)

Resolves #195

DSsoto committed:
THRUSTERS: fix timing dependent thruster detection

* Increase the serial port timeout from 0.01 s to 0.1 s.
* Parse the response to poll packets to confirm that the response packet
  is actually from the thruster we believe it to be.
* Instead of checking for about 100 thruster ID's on each port, only check the
  ones that have been declared on the thruster layout.

kev-the-dev commented on drivers/sub8_videoray_m5_thruster/sub8_thruster_comm/thruster_comm.py:
```diff
@@ -37,10 +36,6 @@ def __init__(self, port_info, thruster_definitions):
              command for ~2 seconds
 
-            - Potential source of problem for future users:
-                If the sub is ever using more than 100 thrusters, and you find that the Sub isn't finding them,
```
SubjuGator9 better have 100 thrusters.

kev-the-dev commented on utils/sub8_diagnostics/scripts/thruster_spinner.py:
```diff
@@ -31,7 +31,10 @@ def ports_from_layout(layout):
             for name in thruster_names:
```
Not relevant to the goal of this PR, but in testing with Ken I found it would be helpful to be able to go reverse in thruster_spinner. I think This just means you have to change the limits on the effort command from [0.0, 1.0] to [-1.0, 1.0]

DSsoto replied:
You're right, I'll add this.

kev-the-dev reviewed (approved):
Seems like a better way to do this, and I assume you tested it so I approve. If the thrusters are plugged in wrong, I think it is better for these to show up as down.

mattlangford commented on utils/sub8_diagnostics/scripts/thruster_spinner.py:
```diff
@@ -31,7 +31,10 @@ def ports_from_layout(layout):
                 port_dict[name] = thruster_port
 
-            msg += "\n\tMotor id's on port: {}".format(thruster_port.motor_ids_on_port)
+            msg += "\n\tMotor id's on port: {}{}{}".format(
```
Assuming you have a printer defined (`p = text_effects.Printer()`), this becomes `msg += str(p.text("Motor id's...").set_bold.blue(np.sort(...)))`
Could be even easier if msg is already a Printer type.

DSsoto replied:
How would you do it if msg was a printer type? Tbh, honest this is harder to use than it needs to be because it doesn't have any docstrings.

mattlangford reviewed (commented):
Consider adding a part to the thruster checker tool that will scan for thrusters and tell you if something's plugged in but not configured (unless this exists and I'm not seeing it)

DSsoto committed:
THRUSTER SPINNER: allow negative effort values

* Also cuts the thrust step in half

DSsoto committed:
THRUSTER COMMS: add check for range of motor ids

* Added a function that takes in a range of motor_ids and returns the following:
  + a list of motor_ids found on that specific port (independent of
    layout config yaml)
  + the average time between the transmission of the poll packet and the
    reception of the response packet for the detected thrusters

DSsoto committed:
THRUSTER SPINNER: add thruster discovery mode
